In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("iphone.csv")   #taking the dataset
print(df.head())

   Gender  Age  Salary  Purchase Iphone
0    Male   19   19000                0
1    Male   35   20000                0
2  Female   26   43000                0
3  Female   27   57000                0
4    Male   19   76000                0


In [3]:
df.isnull().sum() #checking for missing values

Gender             0
Age                0
Salary             0
Purchase Iphone    0
dtype: int64

In [4]:
# Distributing dependent and independent variable

x = df.iloc[:,:-1].values  # x contains independent variable 
y = df.iloc[:,3].values    # y contains dependent variable "Purchase iPhone"


In [5]:
# Converting categorical variable into numbers 

from sklearn.preprocessing import LabelEncoder
le_gender = LabelEncoder()
x[:,0] = le_gender.fit_transform(x[:,0])

In [6]:
# Splitting data into train and test for KNN

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=.25, random_state = 0)  # here splitting testing set to 25%

In [7]:
# Feature Scaling

from sklearn.preprocessing import StandardScaler  # StandardScaler is use to do Feature Transformation with mean = 0 and 
                                                  # std deviation = 1 
sc = StandardScaler()
# In "fit", we are computing the mean and std-deviation in case of StandardScalar; 
# In "transform" we are applying entire formula in each and every data points.
x_train_scalar = sc.fit_transform(x_train)  
x_test_scalar = sc.transform(x_test)  # Data which will use in Making Predictions

In [8]:
# Implementing k-NN Classifier [Model Building]

from sklearn.neighbors import KNeighborsClassifier

# for Euclidean distance we use p = 2 and specify metric as minkowski
c = KNeighborsClassifier(n_neighbors= 5, metric= "minkowski", p =2) # Here, the model will only learn parameters and weight
c.fit(x_train_scalar,y_train)  # Learn from I/P and O/P data and sets its parameter and weight

KNeighborsClassifier()

In [9]:
# Making Predictions

y_pred = c.predict(x_test_scalar)  # It will do the prediction of new data and give us the O/P

In [10]:
from sklearn import metrics

cm = metrics.confusion_matrix(y_test, y_pred) #Creating Confusion Matrix; to compare predicted result with the actual result
print(cm)
accuracy = metrics.accuracy_score(y_test, y_pred) # Checking accuracy of a model
print("Accuracy score:", accuracy)
precision = metrics.precision_score(y_test,y_pred) # Precision score is % of predicted postive events that are actually positive
print("Precision score:", precision)
recall = metrics.recall_score(y_test,y_pred) # Recall score is the % of positive events that we predicted correctly
print("Recall score:", recall)

[[64  4]
 [ 3 29]]
Accuracy score: 0.93
Precision score: 0.8787878787878788
Recall score: 0.90625


<b>Accuracy Score</b> = (TP+TN)/(TP+TN+FP+FN)<br><br>
<b>Recall Score</b> = TP/(TP+FN)<br><br>
<b>Precision Score</b> = TP/(TP+FP)

We get an accuracy of 93%  and only 7 that is 3+4 incorrect predictions were made.<br><br>

64 --> True Negative i.e,; Person has not bought an iPhone and predicted value also says the same<br>
4 --> Flase Positive i.e,; Person has not bought an iPhone but the predicted value says they did buy<br>
3 --> Flase Negative i.e,; Person has bought an iPhone but the predicted value says they did not buy<br>
29 --> True Positive i.e,; Person has bought an iPhone and the predicted value also says they bought<br>